bokehjs/src/lib/models/annotations/arrow.ts

```diff
@@ -56,30 +56,63 @@ export class ArrowView extends DataAnnotationView {
   map_data(): void {
     const {frame} = this.plot_view
 
-    if (this.model.start_units == "data") {
-      this._sx_start = this.coordinates.x_scale.v_compute(this._x_start)
-      this._sy_start = this.coordinates.y_scale.v_compute(this._y_start)
-    } else {
-      this._sx_start = frame.bbox.xview.v_compute(this._x_start)
-      this._sy_start = frame.bbox.yview.v_compute(this._y_start)
-    }
-
-    if (this.model.end_units == "data") {
-      this._sx_end = this.coordinates.x_scale.v_compute(this._x_end)
-      this._sy_end = this.coordinates.y_scale.v_compute(this._y_end)
-    } else {
-      this._sx_end = frame.bbox.xview.v_compute(this._x_end)
-      this._sy_end = frame.bbox.yview.v_compute(this._y_end)
-    }
+    const [sx_start, sy_start] = (() => {
+      switch (this.model.start_units) {
+        case "canvas": {
+          return [
+            new ScreenArray(this._x_start),
+            new ScreenArray(this._y_start),
+          ]
+        }
```

Comment:
so canvas coordinates correspond to a "screen array" but screen coordinates don't? I think we could really use some kind of minimal diagram(s) that visually explains what these different coordinate systems reference and how they are related.

Reply:
"screen" and `ScreenArray` are unrelated. `ScreenArray` is just `Float32Array`. `screen` coordinates are the real issue, because they are erroneously defined with bottom-left corner as origin (relative to renderer's parent), instead of top-left, which makes them really `view` coordinates (they use `{x,y}view.{v_}compute()` afterall). Even if they were actually correctly defined, having true absolute coordinates is beneficial, because we can position renderers without having to re-parent renderers, potentially risking stacking issues or bad clipping. The optimal thing to do for 3.0, would be to actually fix `screen` coordinates, instead of having the current situation when `screen` means top-left origin, except for annotations.

Reply:
Perhaps we could also fix `{x,y}_offset` in `Label*` annotations (`y_offset` is currently subtracted from `sy`).

Reply:
I know they are not related, which is why I think the naming is confusing / unfortunate.